File: server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.druid.server.coordinator.rules;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.common.collect.ImmutableMap;
import org.apache.druid.client.DruidServer;
import org.apache.druid.common.config.Configs;
import org.apache.druid.error.InvalidInput;
import org.apache.druid.timeline.DataSegment;

import java.util.Map;
import java.util.Objects;

/**
 * LoadRules indicate the number of replicants a segment should have in a given tier.
 */
public abstract class LoadRule implements Rule
{
  private final Map<String, Integer> tieredReplicants;
  /**
   * Used to determing the default value if tieredReplicants is null in {@link #handleNullTieredReplicants}.
   */
  private final boolean useDefaultTierForNull;

  private final boolean shouldSegmentBeLoaded;

  protected LoadRule(Map<String, Integer> tieredReplicants, Boolean useDefaultTierForNull)
  {
    this.useDefaultTierForNull = Configs.valueOrDefault(useDefaultTierForNull, true);
    this.tieredReplicants = handleNullTieredReplicants(tieredReplicants, this.useDefaultTierForNull);
    validateTieredReplicants(this.tieredReplicants);
    this.shouldSegmentBeLoaded = this.tieredReplicants.values().stream().reduce(0, Integer::sum) > 0;
  }

  @JsonProperty
  public Map<String, Integer> getTieredReplicants()
  {
    return tieredReplicants;
  }

  @JsonProperty
  public boolean useDefaultTierForNull()
  {
    return useDefaultTierForNull;
  }

  @Override
  public void run(DataSegment segment, SegmentActionHandler handler)
  {
    handler.replicateSegment(segment, getTieredReplicants());
  }


  /**
   * @return Whether a segment that matches this rule needs to be loaded on a tier.
   *
   * Used in making handoff decisions.
   */
  @JsonIgnore
  public boolean shouldMatchingSegmentBeLoaded()
  {
    return shouldSegmentBeLoaded;
  }

  /**
   * Returns the given {@code tieredReplicants} map unchanged if it is non-null (including empty).
   * Returns the following default values if the given map is null.
   * <ul>
   * <li>If {@code useDefaultTierForNull} is true, returns a singleton map from {@link DruidServer#DEFAULT_TIER} to {@link DruidServer#DEFAULT_NUM_REPLICANTS}.</li>
   * <li>If {@code useDefaultTierForNull} is false, returns an empty map. This causes segments to have a replication factor of 0 and not get assigned to any historical.</li>
   * </ul>
   */
  private static Map<String, Integer> handleNullTieredReplicants(
      final Map<String, Integer> tieredReplicants,
      boolean useDefaultTierForNull
  )
  {
    if (useDefaultTierForNull) {
      return Configs.valueOrDefault(
          tieredReplicants,
          ImmutableMap.of(DruidServer.DEFAULT_TIER, DruidServer.DEFAULT_NUM_REPLICANTS)
      );
    } else {
      return Configs.valueOrDefault(tieredReplicants, ImmutableMap.of());
    }
  }

  private static void validateTieredReplicants(final Map<String, Integer> tieredReplicants)
  {
    for (Map.Entry<String, Integer> entry : tieredReplicants.entrySet()) {
      if (entry.getValue() == null) {
        throw InvalidInput.exception(
            "Invalid number of replicas for tier [%s]. Value must not be null.",
            entry.getKey()
        );
      }
      if (entry.getValue() < 0) {
        throw InvalidInput.exception(
            "Invalid number of replicas for tier [%s]. Value [%d] must be positive.",
            entry.getKey(),
            entry.getValue()
        );
      }
    }
  }

  public int getNumReplicants(String tier)
  {
    Integer retVal = getTieredReplicants().get(tier);
    return (retVal == null) ? 0 : retVal;
  }

  @Override
  public boolean equals(Object o)
  {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    LoadRule loadRule = (LoadRule) o;
    return useDefaultTierForNull == loadRule.useDefaultTierForNull && Objects.equals(
        tieredReplicants,
        loadRule.tieredReplicants
    );
  }

  @Override
  public int hashCode()
  {
    return Objects.hash(tieredReplicants, useDefaultTierForNull);
  }
}

File: web-console/src/utils/load-rule.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { sum } from 'd3-array';

import { deepMove, deepSet } from './object-change';

export type RuleType =
  | 'loadForever'
  | 'loadByInterval'
  | 'loadByPeriod'
  | 'dropForever'
  | 'dropByInterval'
  | 'dropByPeriod'
  | 'dropBeforeByPeriod'
  | 'broadcastForever'
  | 'broadcastByInterval'
  | 'broadcastByPeriod';

export interface Rule {
  type: RuleType;
  interval?: string;
  period?: string;
  includeFuture?: boolean;
  tieredReplicants?: Record<string, number>;
}

export class RuleUtil {
  static TYPES: RuleType[] = [
    'loadForever',
    'loadByInterval',
    'loadByPeriod',
    'dropForever',
    'dropByInterval',
    'dropByPeriod',
    'dropBeforeByPeriod',
    'broadcastForever',
    'broadcastByInterval',
    'broadcastByPeriod',
  ];

  static ruleToString(rule: Rule): string {
    const params: string[] = [];

    if (RuleUtil.hasPeriod(rule)) {
      params.push(`${rule.period}${RuleUtil.getIncludeFuture(rule) ? '+future' : ''}`);
    }
    if (RuleUtil.hasInterval(rule)) params.push(rule.interval || '?');
    if (RuleUtil.canHaveTieredReplicants(rule)) params.push(`${RuleUtil.totalReplicas(rule)}x`);

    return `${rule.type}(${params.join(', ')})`;
  }

  static changeRuleType(rule: Rule, type: RuleType): Rule {
    const newRule = deepSet(rule, 'type', type);

    if (RuleUtil.hasPeriod(newRule)) {
      newRule.period ??= 'P1M';
      newRule.includeFuture ??= true;
    } else {
      delete newRule.period;
      delete newRule.includeFuture;
    }

    if (RuleUtil.hasInterval(newRule)) {
      newRule.interval ??= '2010-01-01/2020-01-01';
    } else {
      delete newRule.interval;
    }

    if (RuleUtil.canHaveTieredReplicants(newRule)) {
      newRule.tieredReplicants ??= { _default_tier: 2 };
    } else {
      delete newRule.tieredReplicants;
    }

    return newRule;
  }

  static hasPeriod(rule: Rule): boolean {
    return rule.type.endsWith('ByPeriod');
  }

  static changePeriod(rule: Rule, period: string): Rule {
    return deepSet(rule, 'period', period);
  }

  static hasIncludeFuture(rule: Rule): boolean {
    return RuleUtil.hasPeriod(rule) && rule.type !== 'dropBeforeByPeriod';
  }

  static getIncludeFuture(rule: Rule): boolean {
    return rule.includeFuture ?? true;
  }

  static changeIncludeFuture(rule: Rule, includeFuture: boolean): Rule {
    return deepSet(rule, 'includeFuture', includeFuture);
  }

  static hasInterval(rule: Rule): boolean {
    return rule.type.endsWith('ByInterval');
  }

  static changeInterval(rule: Rule, interval: string): Rule {
    return deepSet(rule, 'interval', interval);
  }

  static canHaveTieredReplicants(rule: Rule): boolean {
    return rule.type.startsWith('load');
  }

  static renameTieredReplicants(rule: Rule, oldTier: string, newTier: string): Rule {
    return deepMove(rule, `tieredReplicants.${oldTier}`, `tieredReplicants.${newTier}`);
  }

  static addTieredReplicant(rule: Rule, tier: string, replication: number): Rule {
    const newTieredReplicants = deepSet(rule.tieredReplicants || {}, tier, replication);
    return deepSet(rule, 'tieredReplicants', newTieredReplicants);
  }

  static totalReplicas(rule: Rule): number {
    return sum(Object.values(rule.tieredReplicants || {}));
  }

  static isZeroReplicaRule(rule: Rule): boolean {
    return RuleUtil.canHaveTieredReplicants(rule) && RuleUtil.totalReplicas(rule) === 0;
  }

  static hasZeroReplicaRule(rules: Rule[] | undefined, defaultRules: Rule[] | undefined): boolean {
    return (
      (rules || []).some(RuleUtil.isZeroReplicaRule) ||
      (defaultRules || []).some(RuleUtil.isZeroReplicaRule)
    );
  }
}
